Speedup some picotool commands

will-v-pi opened the pull request:
This offers substantial performance improvements for `info` (with no signing/hashing) and `save`/`verify`, and small improvements for `load`. The performance figures for a 4MiB binary on Pico 2 are below, and on Pico the figures are also improved (although with less of an improvement for `info` due to the Pico not including metadata blocks). This fixes #204.

| Command | speedups | 2.1.1 | 2.1.0 | 2.0.0 |
|--------|--------|--------|--------|--------|
| `info` | 0.056 | 9.156 | 18.488 | 9.346 |
| `info` (hashed binary) | 9.156 | 9.173 | 36.747 | 18.408 |
| `save` | 9.8 | 21.747 | 31.28 | 35.869 |
| `save --range` | 9.744 | 40.058 | 23.325 | 26.087 |
| `load` | 65.121 | 70.688 | 72.645 | 89.636 |

The main fixes have been:
- Batching reads/writes into >4096 byte chunks, rather than the existing smaller chunk sizes (eg 256 byte chunks for `save`)
- Doing a sparse read of the block loop when not verifying the binary hash and only read the whole binary when verifying the hash, rather than always reading the whole binary in case the hash needs to be verified.

Also includes 3 drive-by fixes:
- Fixed a bug with truncated binaries, where the metadata search would get into an infinite loop - this fix was already in `get_last_block` but was missing from `get_all_blocks`
- Replace `<unknown>` printout for `PICOBOOT_UNKNOWN_ERROR` with `unknown error`, to clarify the error was not an unknown error value, just an unknown error - the `<unknown>` printout is still used for other unknown error values
- Workaround picoboot interface not supporting >4MiB reads - this only crops up when validating the hash/signature for >4MiB binaries, as it reads the whole binary in one go

will-v-pi committed:
Only use flash caching for `info`

It only offers a performance improvement when the same flash is read multiple times, such as with block searches in `info`
It can degrade performance in other situations such as `save` which have no need for a cache

will-v-pi committed:
Allow piecemeal reads during block searches

For unsigned/unhashed binaries, this offers a large performance boost, by only reading the metadata blocks and not reading the whole binary.
For signed/hashed binaries there is no difference, as the whole binary must be read for verification anyway.

will-v-pi committed:
Fix bug with truncated binaries

Without this you get an infinite loop - this fix was already in get_last_block

will-v-pi committed:
Read in larger chunks when saving

This speeds up picotool save by reading in larger chunk sizes, with a chunk size of 4096 or size/100, whichever is greater

will-v-pi committed:
Allow batching of loads

This offers a small performance boost, but there is room for more of a boost at the expense of the progress bar smoothness

will-v-pi committed:
Use larger batches when verifying too

will-v-pi committed:
Add "unknown error" printout for `PICOBOOT_UNKNOWN_ERROR`

This is instead of printing <unknown>, which could be any unknown error value

will-v-pi committed:
Workaround picoboot interface not supporting reads over 4MiB

lurch commented on main.cpp:
```diff
@@ -1867,7 +1868,7 @@ struct picoboot_memory_access : public memory_access {
 
     void read(uint32_t address, uint8_t *buffer, unsigned int size, __unused bool zero_fill) override {
-        if (flash == get_memory_type(address, model)) {
+        if (flash == get_memory_type(address, model) && settings.use_flash_cache) {
```
Is it worth swapping these conditionals around, as I guess if `settings.use_flash_cache` returns false then there's no need to check `flash == get_memory_type(address, model)` ?

lurch commented on bintool/bintool.cpp:
```diff
@@ -362,12 +365,15 @@ std::vector<std::unique_ptr<block>> get_all_blocks(std::vector<uint8_t> &bin, ui
-        auto offset = next_block_addr - storage_addr;
-        if (offset + read_size > bin.size() && more_cb != nullptr) {
-            more_cb(bin, offset + read_size);
+        if (next_block_addr + read_size > current_bin_start + bin.size() && more_cb != nullptr) {
```
The change here (lines 368-376) looks identical to the previous change (on lines 322-329). Does that suggest that there's some common code that might be worth pulling out into a separate function? :shrug:

will-v-pi replied:
They had entirely common code, so I've changed `get_last_block` to now call `get_all_blocks` and just return the last block.

will-v-pi committed:
Review fixups

Re-order flash caching `if`

Make `get_last_block` just call `get_all_blocks` as there shouldn't be a performance hit (the only difference is `get_all_blocks` stores copies of all the blocks, whereas `get_last_block` only stored a copy of the latest block)

kilograham commented on bintool/bintool.cpp:
```diff
@@ -315,60 +315,23 @@ block place_new_block(elf_file *elf, std::unique_ptr<block> &first_block) {
-        auto offset = next_block_addr - storage_addr;
-        if (offset + read_size > bin.size() && more_cb != nullptr) {
-            more_cb(bin, offset + read_size);
+        if (next_block_addr + read_size > current_bin_start + bin.size() && more_cb != nullptr) {
```
seems like we should track which blocks we've already seen to catch loops or is that done elsewhere?

will-v-pi replied:
Valid block loops are caught below - `Found last block in block loop`

Lollipop block loops are not caught so will (probably) loop forever, but they are invalid anyway, and will trigger E22

kilograham commented on main.cpp:
```diff
@@ -4303,6 +4308,7 @@ bool save_command::execute(device_map &devices) {
         fail(ERROR_NOT_POSSIBLE, "Save range crosses unmapped memory");
     }
     uint32_t size = end - start;
+    uint32_t chunk_size = std::max(size / 100, FLASH_SECTOR_ERASE_SIZE);
```
I think i'd prefer chunk_size / 100 rounded up to the next FLASH_SECTOR_ERASE_SIZE multiple

kilograham replied:
not critical for read, but perhaps tidier and more consistent with others

kilograham commented on main.cpp:
```diff
@@ -4592,7 +4602,9 @@ bool load_guts(picoboot::connection con, iostream_memory_access &file_access) {
             progress_bar bar("Loading into " + memory_names[type] + ": ");
-            uint32_t batch_size = FLASH_SECTOR_ERASE_SIZE;
+            // Use batches of size/100 rounded up to FLASH_SECTOR_ERASE_SIZE
+            uint32_t batch_size = std::max(FLASH_SECTOR_ERASE_SIZE,
```
arguably the std::max is unnecessary, since everything but 0 would be rounded up anyway

kilograham commented on main.cpp:
```diff
@@ -4637,7 +4649,7 @@ bool load_guts(picoboot::connection con, iostream_memory_access &file_access) {
             {
                 progress_bar bar("Verifying " + memory_names[type] + ": ");
-                uint32_t batch_size = FLASH_SECTOR_ERASE_SIZE;
+                uint32_t batch_size = std::max(mem_range.len() / 100, FLASH_SECTOR_ERASE_SIZE);
```
ditto batch size - perhaps make a function

kilograham commented on main.cpp:
```diff
@@ -5272,7 +5284,7 @@ bool verify_command::execute(device_map &devices) {
                     vector<uint8_t> file_buf;
                     vector<uint8_t> device_buf;
-                    uint32_t batch_size = 1024;
+                    uint32_t batch_size = std::max(mem_range.len() / 100, FLASH_SECTOR_ERASE_SIZE);
```
ditto

will-v-pi committed:
Add common `calculate_chunk_size` function

This calculates chunk_size as size/100 rounded up to FLASH_SECTOR_ERASE_SIZE
